app/pages/lab-pages-editor/DataManager.jsx

```diff
@@ -0,0 +1,97 @@
+/*
+Data Manager
+Responsible for fetching, updating, and saving Workflow resources from/to
+the Panoptes service.
+ */
+
+import React, { useEffect, useMemo, useState } from 'react';
+import PropTypes from 'prop-types';
+import apiClient from 'panoptes-client/lib/api-client';
+import { WorkflowContext } from './context.jsx';
+
+function DataManager({
+  // key: to ensure DataManager renders FRESH (with states reset) whenever workflowId changes, use <DataManager key={workflowId} ... />
+  children,
+  workflowId
+}) {
+  let initialised = false;
+  const [apiData, setApiData] = useState({
+    workflow: null,
+    status: 'ready'
+  });
+
+  async function fetchWorkflow() {
+    try {
+      setApiData({
+        workflow: null,
+        status: 'fetching'
+      });
+
+      const wf = await apiClient.type('workflows').get(workflowId);
+      if (!wf) throw new Error('No workflow');
+
+      setApiData({
+        workflow: wf,
+        status: 'ready'
+      });
+    } catch (err) {
+      console.error('DataManager: ', err);
+      setApiData({
+        workflow: null,
+        status: 'error'
+      });
+    }
+  }
+
+  // Fetch workflow when the component loads for the first time.
+  // See notes about 'key' prop, to ensure states are reset:
+  // https://react.dev/learn/you-might-not-need-an-effect#resetting-all-state-when-a-prop-changes
+  // Also see general pattern notes:
+  // https://react.dev/reference/react/useEffect#fetching-data-with-effects
+  useEffect(() => {
+    if (!initialised) fetchWorkflow();
+    initialised = true;
+    return () => { initialised = false; };
+  }, [workflowId]);
+
+  /*
+  Updates the workflow with new data.
+   */
+  function update(data) {
+    console.log('+++ TODO');
+  }
+
+  // Wrap contextData in a memo so it doesn't re-create a new object on every render.
+  // See https://react.dev/reference/react/useContext#optimizing-re-renders-when-passing-objects-and-functions
+  const contextData = useMemo(() => ({
+    workflow: apiData.workflow,
+    update
+  }), [apiData.workflow]);
```

Comment:
```suggestion
  }), [apiData.workflow, update]);
```

`update` is also an external dependency here. `update` is also declared on every render, which means your context data is a new object on every render.

Reply:
Noted, I'll need to figure out where to house update() once I find the final home for the data fetch functionality.

Reply:
Your code here will work, without React errors, if you declare update inside the `contextData` object:

```js
const contextData = useMemo(() => ({
    workflow: apiData.workflow,
    update: () => {
      console.log('update the workflow here')
    }
  }), [apiData.workflow]);
```

Call `contextData.update()`, rather than `update()`, in order to invoke it.

Otherwise, you will have to declare `update` as a dependency here, and likely also wrap `update` in `useCallback`. Like `useEffect`, you can't choose dependencies for `useMemo` or `useCallback`. You must declare all external reactive values used inside the hook. See the `login` function in the [useContext example code](https://react.dev/reference/react/useContext#optimizing-re-renders-when-passing-objects-and-functions).